pkg/prom/http_test.go

```diff
@@ -39,3 +44,64 @@ func TestAgent_ListInstancesHandler(t *testing.T) {
 		})
 	})
 }
+
+func TestAgent_ListTargetsHandler(t *testing.T) {
+	fact := newMockInstanceFactory()
+	a, err := newAgent(Config{
+		WALDir: "/tmp/agent",
+	}, log.NewNopLogger(), fact.factory)
+	require.NoError(t, err)
+
+	r := httptest.NewRequest("GET", "/agent/api/v1/targets", nil)
+
+	t.Run("scrape manager not ready", func(t *testing.T) {
+		a.instances = map[string]inst{
+			"test_instance": &mockInstanceScrape{},
+		}
+
+		rr := httptest.NewRecorder()
+		a.ListTargetsHandler(rr, r)
+		expect := `{"status":"success","data":[]}`
+		require.Equal(t, expect, rr.Body.String())
+	})
+
+	t.Run("scrape manager targets", func(t *testing.T) {
+		tgt := scrape.NewTarget(labels.FromMap(map[string]string{
+			model.JobLabel:         "job",
+			model.InstanceLabel:    "instance",
+			model.SchemeLabel:      "http",
+			model.AddressLabel:     "localhost:12345",
+			model.MetricsPathLabel: "/metrics",
+			"foo":                  "bar",
+		}), nil, nil)
+
+		startTime := time.Date(1994, time.January, 12, 0, 0, 0, 0, time.UTC)
+		tgt.Report(startTime, time.Minute, fmt.Errorf("something went wrong"))
+
+		a.instances = map[string]inst{
+			"test_instance": &mockInstanceScrape{
+				tgts: map[string][]*scrape.Target{
+					"group_a": {tgt},
+				},
+			},
+		}
+
+		rr := httptest.NewRecorder()
+		a.ListTargetsHandler(rr, r)
+		expect := `{"status":"success","data":[{"instance":"test_instance","target_group":"group_a","endpoint":"http://localhost:12345/metrics","state":"down","labels":{"foo":"bar","instance":"instance","job":"job"},"last_scrape":"1994-01-12T00:00:00Z","scrape_duration_ms":60000,"scrape_error":"something went wrong"}]}`
```

Comment:
nitpick question: Is this preferable to building up an `expect` struct and unmarshalling rr.Body to compare? It would make this test less brittle to changes to `TargetInfo`

Reply:
it's a good question. @gotjosh and I had talked about this a while ago, where he said he generally prefers testing for actual string content. WDYT @gotjosh, any more detail on one vs the other? 

I'm happy to change this either way, and I acknowledge the expectation string is pretty unwieldy.

Reply:
I'm personally fond of just using the string version (the only thing I'd suggest is that we actually intent it).

I find that is very simple to evolve as you don't need to make struct changes across multiple places, and it reflects accurately what one should see when you test it. Otherwise, you end up making struct changes and following the JSON definition of them around the codebase before you figure out how you want this thing to look like.

Plus, if you use something like https://godoc.org/github.com/stretchr/testify/assert#JSONEq then the diff is👌   which tends to be my biggest worry.

Reply:
Oh wow, I didn't know about JSONEq and that fixes a lot of my problems here. I'll change the tests to use that instead and break up the giant string into a more readable string.